### DISCUSSION TITLE
Details about `ZeroCopyBuffer`

### DISCUSSION THREAD

**Discussion by temeddix** [fzyzcjy/flutter_rust_bridge#1221] [Q&A] [answered]
Hi, I have some new questions.

1. Can I make the function's return type to be something like `ZeroCopyBuffer<MyCustomStruct>`? Would this work as intended, without copying the fields of `MyCustomStruct`?

2. Is there a way to pass bytes to Dart, without letting go of the ownership in Rust? I wonder if there would be some data type like `Arc`, that can be used both in Dart and Rust, without any copying.
    ```rust
    pub fn my_function() -> Vec<u8> {
        let bytes = GLOBAL_HASHMAP.get("someKey").unwrap();
        bytes
    }
    ```
    In this case, we give away the ownership of `bytes` and are not able to access it anymore in Rust.
    ```rust
    pub fn my_function() -> ZeroCopyBuffer<Vec<u8>> {
        let bytes = GLOBAL_HASHMAP.get("someKey").unwrap();
        ZeroCopyBuffer(bytes)
    }
    ```
    Even if we use `ZeroCopyBuffer`, we cannot access that data in Rust anymore.

**Comment by fzyzcjy** [comment, marked as answer]
> Can I make the function's return type to be something like ZeroCopyBuffer<MyCustomStruct>? Would this work as intended, without copying the fields of MyCustomStruct?

Yes and no.

ZeroCopyBuffer is implemented by Dart primitives, which only support a flat byte array.

But, if you generate some Dart code like:

```
class Apple {
  final Uint8List bytes;
  int get aaa => bytes[0] * 256 + bytes[1];
  int get bbb => bytes[2];
}
```

Then I guess it should work.

> I wonder if there would be some data type like Arc, that can be used both in Dart and Rust, without any copying.

I am worried that this may violate the memory model of Rust very easily. For example, you can modify the bytes in Dart and Rust at the same time.

However, there can be fancier approaches.

Possibly related: https://github.com/fzyzcjy/flutter_rust_bridge/issues/1132

**Comment by temeddix** [comment]
Okay, thank you very much for the help!

**Reply by fzyzcjy** [reply]
You are welcome!
